File: Week_03/G20190343020039/LeetCode_22_0039.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  let result = [];
  recursion(n, 0, 0, "", result);
  return result;
};

function recursion(n, left, right, cur, result) {
  if (left === n && right === n) {
    result.push(cur);
    return;
  }

  if (left < n) {
    recursion(n, left + 1, right, cur + "(", result);
  }

  if (right < left) {
    recursion(n, left, right + 1, cur + ")", result);
  }
}

File: Week_06/G20190343020039/LeetCode_22_0039.py
class Solution:

  def generateParenthesis(self, n: int) -> List[str]:
    result = []
    self.dfs(n, 0, 0, "", result)
    return result

  def dfs(self, n, left, right, node, result):
    if left == n and right == n:
      result.append(node)
    if left < n:
      self.dfs(n, left + 1, right, node + "(", result)
    if right < left:
      self.dfs(n, left, right + 1, node + ")", result)

